### PR TITLE
Aidex support broadcast

### PR DESCRIPTION
**Pull request by andyrozman**
This is code for supporting Aidex Broadcasts (from their application). 

They have released app version that supports broadcast worldwide few weeks ago, but it has one small bug inside, so fixed version should be coming with their next version.

### CODE REVIEW

**Comment by jamorham** [comment]
Can you describe the testing which has occurred so far with this?

Please also confirm the copyright status of the supplied logo image.

Thanks.

**Comment by andyrozman** [comment]
The Logo was not supplied by me, so I can't comment on that...

Testing that was done, was just simple. Using both apps together (I tested this with pre-released version then) and seeing that data displayed is the same. 

Testing with released app was done by other user (I don't have Aidex sensors atm) and he detected that their app has a problem, which is visible in Xdrip as deltas not being displayed.

File: app/src/main/java/com/eveningoutpost/dexdrip/receivers/aidex/AidexReceiver.java
Comment (andyrozman) [reply app/src/main/java/com/eveningoutpost/dexdrip/receivers/aidex/AidexReceiver.java:210]: You are right. My knowledge of Xdrip is limited, and I found this code, and I thought this is the way to do it. 
But you are right, we get already processed data.

Please let me know how to fix it and I will do it.

**Comment by jamorham** [comment]
Merging a graphic in to the project source tree effectively places it under the GPL. 

Everything submitted for inclusion to the project needs to be either under the same or more permissive license or be the authors original work so that it doesn't taint the copyright status of the project.

It is possible to include copyrighted images, but I would prefer not to do that as (afaik) we would need a separate license and exclusion notice for it which xDrip has so far avoided having to do. 

We would need the copyright holder of the image to agree that it is in the public domain or that they are assigning copyright to the project.  

At this stage it might just be easier to remove the graphic because I wouldn't normally add a new collector type to the source wizard until it gains some traction anyway.

**Comment by andyrozman** [comment]
Sure. Like I said, I didn't add the image. 

You know your product best, so if you say remove it I will remove it...

**Comment by andyrozman** [comment]
Do you have generic picture I could use here?

File: app/src/main/java/com/eveningoutpost/dexdrip/receivers/aidex/AidexReceiver.java
Comment (tolot27) [inline app/src/main/java/com/eveningoutpost/dexdrip/receivers/aidex/AidexReceiver.java:181]: Here you can use `AidexBroadcastIntents.UNIT_MG_DL` as well.

**Review by tolot27 — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/eveningoutpost/dexdrip/receivers/aidex/AidexReceiver.java
Comment (andyrozman) [reply app/src/main/java/com/eveningoutpost/dexdrip/receivers/aidex/AidexReceiver.java:83]: Depends on what is happening... In this case they are not necessary. Will remove.

File: app/src/main/java/com/eveningoutpost/dexdrip/receivers/aidex/AidexReceiver.java
Comment (andyrozman) [reply app/src/main/java/com/eveningoutpost/dexdrip/receivers/aidex/AidexReceiver.java:181]: Fixed.

File: app/src/main/java/com/eveningoutpost/dexdrip/Models/Sensor.java
Comment (jamorham) [inline app/src/main/java/com/eveningoutpost/dexdrip/Models/Sensor.java:161]: Would this function get repeatedly triggered by the buggy version of manufacturer app?

If so can you wrap it with a rate limit using `pratelimit` as it initiates network activity potentially and we don't want a buggy app to be able to cause that problem internally.

**Comment by jamorham** [comment]
I suggest removing the logo and the source wizard functionality for the moment.

File: app/src/main/java/com/eveningoutpost/dexdrip/receivers/aidex/AidexReceiver.java
Comment (jamorham) [reply app/src/main/java/com/eveningoutpost/dexdrip/receivers/aidex/AidexReceiver.java:210]: You can do like ` BloodTest.create(timestamp, mgdl, "Aidex", null)` - Timestamp is milliseconds since epoch, for now use `JoH.tsl()` and mgdl is glucose in mgdl.

Comment (andyrozman) [reply app/src/main/java/com/eveningoutpost/dexdrip/receivers/aidex/AidexReceiver.java]: fixed.

**Comment by andyrozman** [comment]
> I suggest removing the logo and the source wizard functionality for the moment.

How will user be able to select Aidex option if its not it that wizard?

**Comment by Navid200** [comment]
I wanted to put together a guide for graphics because this has come up before, and ask developers to approve.  But, I have difficulty uploading graphics to github.

My suggestion is to create graphics only containing text and I provided a methodology to allow us to follow a standard.  
But, I cannot complete the page because I cannot upload.  Hopefully, it will be resolved.

@jamorham
Will this be acceptable?
aidex_icon

It's a png file with a 150x113 size, the same as the xDrip Sync follower icon.

File: app/src/main/res/values/strings.xml
Comment (andyrozman) [reply app/src/main/res/values/strings.xml]: I removed the strings, since we are not using them at the moment.

**Comment by andyrozman** [comment]
> I suggest removing the logo and the source wizard functionality for the moment.

I used that icon supplied by Navid200.

**Comment by Navid200** [comment]
@andyrozman I'm sorry if my post was not clear.
That was a question for @jamorham to approve.  He may not.

File: app/src/main/java/com/eveningoutpost/dexdrip/Models/Sensor.java
Comment (jamorham) [reply app/src/main/java/com/eveningoutpost/dexdrip/Models/Sensor.java:161]: Why are we restarting the sensor every 5 minutes?

**Comment by jamorham** [comment]
Please refer to my previous comment about the Source Wizard graphic. 
https://github.com/NightscoutFoundation/xDrip/pull/2312#issuecomment-1240803034

**Comment by andyrozman** [comment]
> Please refer to my previous comment about the Source Wizard graphic. [#2312 (comment)](https://github.com/NightscoutFoundation/xDrip/pull/2312#issuecomment-1240803034)

Commented out.

File: app/src/main/java/com/eveningoutpost/dexdrip/Models/Sensor.java
Comment (andyrozman) [reply app/src/main/java/com/eveningoutpost/dexdrip/Models/Sensor.java:161]: We are not, but they are sending us different sensorId with each broadcast (which is error). We could wait for their fix and then merge...

File: app/src/main/java/com/eveningoutpost/dexdrip/Home.java
Comment (jamorham) [inline app/src/main/java/com/eveningoutpost/dexdrip/Home.java:834]: I'm not seeing this used in your code? Is that correct? If so please revert the visibility of this method. Thanks

**Comment by jamorham** [comment]
I would like to merge this PR but it still needs some simple changes.
- Please resolve the merge conflict in `AndroidManifest.xml` - likely caused by reformatting changes from another PR
- Please remove the unused logo image
- If the `startIntentThreadWithDelayedRefresh` method is not used by your code please revert the visibility of this
- I believe that the sensor restarting method causes network activity because it will try to sync the new sensor information. As such this needs to be rate limited so that if the aidex application misbehaves this cannot cause problems outside of the local xDrip instance. You can use a `pratelimit` call for this - see examples in the code. This is, from my understanding, especially important as there is already a published version of the aidex application which will trigger this failure condition.

With those changes in place I would expect to be able to merge this PR. I have updated xDrip's contributing document to make it clear about copyright status of submissions including graphics etc and for this PR I would plan to squash the merge to eliminate any questionable copyrighted graphics from the repository.

**Comment by jamorham** [comment]
Please let me know if there are plans to follow the requested changes. Thanks.

**Comment by andyrozman** [comment]
I thought that we were waiting for them to fix their app first... 

I can fix this problems over weekend... and then we can probably merge it...

**Comment by andyrozman** [comment]
I fixed all problems you mentioned @jamorham

**Comment by andyrozman** [comment]
@jamorham Is there anything left I need to do for this?

File: app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/SourceWizard.java
Comment (tolot27) [inline app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/SourceWizard.java:76]: Please remove unnecessary lines.

File: app/src/main/java/com/eveningoutpost/dexdrip/receivers/aidex/AidexReceiver.java
Comment (tolot27) [inline app/src/main/java/com/eveningoutpost/dexdrip/receivers/aidex/AidexReceiver.java:59]: Please remove commented out lines.

**Review by tolot27 — APPROVED** [review]
Looks good so far. Just a few small code cleanups are necessary.

File: app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/SourceWizard.java
Comment (andyrozman) [reply app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/SourceWizard.java]: fixed

File: app/src/main/java/com/eveningoutpost/dexdrip/receivers/aidex/AidexReceiver.java
Comment (andyrozman) [reply app/src/main/java/com/eveningoutpost/dexdrip/receivers/aidex/AidexReceiver.java]: fixed

File: app/src/main/java/com/eveningoutpost/dexdrip/Home.java
Comment (tolot27) [reply app/src/main/java/com/eveningoutpost/dexdrip/Home.java]: That's reverted.

File: app/src/main/java/com/eveningoutpost/dexdrip/Models/Sensor.java
Comment (tolot27) [reply app/src/main/java/com/eveningoutpost/dexdrip/Models/Sensor.java:161]: `pratelimit` is now used.

**Review by tolot27 — APPROVED** [review]
@jamorham IMHO, it is ready for merging.

**Comment by Stealthwick** [comment]
I have tested this code and it appears to work well. It would be useful for me and others to have this merged. Any chance we could pull the trigger on this?

**Comment by jamorham** [comment]
This still doesn't have the rate limiting on sensor restarting but I can adopt this PR and make the changes myself if that would help.

**Comment by tolot27** [comment]
> This still doesn't have the rate limiting on sensor restarting but I can adopt this PR and make the changes myself if that would help.

I believe ef438ad44 adds it.

File: app/src/main/java/com/eveningoutpost/dexdrip/receivers/aidex/AidexReceiver.java
Comment (jamorham) [inline app/src/main/java/com/eveningoutpost/dexdrip/receivers/aidex/AidexReceiver.java:236]: @tolot27 this is the function I was referring to.

**Comment by jamorham** [comment]
The function still requiring rate-limiting is `processSensorRestart()` I don't believe this has been addressed looking at the PR.

File: app/src/main/java/com/eveningoutpost/dexdrip/receivers/aidex/AidexReceiver.java
Comment (tolot27) [inline app/src/main/java/com/eveningoutpost/dexdrip/receivers/aidex/AidexReceiver.java:164]: It looks like code formatting isn't applied.

File: app/src/main/java/com/eveningoutpost/dexdrip/receivers/aidex/AidexReceiver.java
Comment (tolot27) [reply app/src/main/java/com/eveningoutpost/dexdrip/receivers/aidex/AidexReceiver.java:236]: Is rate limiting required in `checkIfCorrectSensorIsRunning`? Maybe, @andyrozman misunderstood the place where to implement it.

Comment (andyrozman) [reply app/src/main/java/com/eveningoutpost/dexdrip/receivers/aidex/AidexReceiver.java:236]: Actually processSensorRestart would be called only if we received this event from their side, which *might* happen when they restart sensor (I am not sure if they implemented it that way)... But method checkIfCorrectSensorIsRunning() is called on each New_BG event, which is sent every 5 minutes. 
So checkIfCorrectSensorIsRunning() is the one that must be rate limited...

**Comment by jamorham** [comment]
I've merged this and added the additional rate limiting safety check:
https://github.com/NightscoutFoundation/xDrip/commit/3e2ec7593231d7a30d2f46affacfb1d06203f86d?diff=unified#diff-7dec58fa33a653e7d3df6887f7265e93adb9c96dfe2a3163aa055159704406a0R233

As we know the existing manufacturer app can be buggy with how it presents sensor identity, we need to ensure that it can't break anything or create undue network traffic if it does something crazy which is what this addition is designed to prevent.

**Pull request merged**
